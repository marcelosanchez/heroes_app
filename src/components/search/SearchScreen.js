import React, { useMemo } from 'react'
import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {
    const navigate = useNavigate();  // permite navegar entre las rutas
    const location = useLocation();  // permite obtener la ruta actual

    const { q = '' } = queryString.parse( location.search );  // obtiene los parametros de la ruta


    const [ formValues, handleInputChange ] = useForm( {  // uso el custom hook
        searchText: q,  // valor por defecto
    } );

    const { searchText } = formValues;  // desestructurando

    const heroesFiltered = useMemo( () => {
        return getHeroesByName( q );
    } , [ q ] );  // memoiza el resultado

    const handleSearch = (e) => {
        e.preventDefault(); 
        navigate(`?q=${ searchText }`);
    }

    return (
        <>
            <h1>Search Results</h1>
            <hr />

            <div className='row'>

                <div className='col-lg-5'>
                    <h4>Search</h4>
                    <hr/>
                    <form onSubmit={ handleSearch }>
                        <input
                            type='text'
                            placeholder='Search hero'
                            className='form-control'
                            name='searchText'
                            autoComplete='off'
                            value={ searchText }
                            onChange={ handleInputChange }
                        />

                        <button 
                            className='btn btn-outline-primary mt-1'
                            type='submit' 
                        > 
                            Search
                        </button>
                    </form>
                </div>

                <div className='col-lg-7'>
                    <h4>Results</h4>
                    <hr/>

                    {
                        (q === '') 
                            ? <div className='alert alert-info'>Please enter a search term</div>
                            : ( heroesFiltered.length === 0 )
                                && <div className='alert alert-danger'>No results found: { q }</div>
                    }   


                    <ul className='list-group'>
                        {
                            heroesFiltered.map( hero => (
                                <HeroCard 
                                    key={ hero.id }
                                    { ...hero }
                                />
                            ) )
                        }
                    </ul>
                </div>

            </div>

        </>
    )
}
