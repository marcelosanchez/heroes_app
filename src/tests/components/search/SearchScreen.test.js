import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';

const mockNavigate = jest.fn();  // mock de la funcion navigate

jest.mock('react-router-dom', () => ({  // mock de react-router-dom
    ...jest.requireActual('react-router-dom'),  // importa todo lo que hay en react-router-dom
    useNavigate: () => mockNavigate,  // modifica la funcion navigate para llamarla en el test
}))

describe('Pruebas en <SearchScreen />', () => {
    test('Deberia mostrarse correctamente', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-info').text().trim() ).toBe('Please enter a search term');
    })

    test('Deberia mostrar a Batman y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('input').prop('value') ).toBe('batman');
    });

    test('Deberia mostrar un error si no se encuentra el heroe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search?q=batman123'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.alert-danger').text().trim() ).toBe('No results found: batman123');
        // Tarea
        // No hay resultados: batman123
    });

    test('Debe llamar el navigate a la nueva URL', () => {


        const wrapper = mount(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {  // simulo el evento change del input
            target: {  // simulo el target del evento
                name: 'searchText',  // nombre del input
                value: 'batman',  // valor del input
            }
        });

        wrapper.find('form').prop('onSubmit')({  // simulo el evento submit del form
            preventDefault: () => {},  // simulo el preventDefault del evento
        });

        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');  // verifico que se llame el navigate
    });
})