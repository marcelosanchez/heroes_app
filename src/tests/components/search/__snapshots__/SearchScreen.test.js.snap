// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Deberia mostrar a Batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search Results
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <ul
        className="list-group"
      >
        <div
          className="col animate__animate animate__fadeIn"
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4"
              >
                <img
                  alt="Batman"
                  className="card-img"
                  src="./assets/dc-batman.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="card-body"
                >
                  <h5
                    className="card-title"
                  >
                    Batman
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Wayne
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Detective Comics #27
                    </small>
                  </p>
                  <a
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    More info...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </ul>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> Deberia mostrar un error si no se encuentra el heroe 1`] = `
Array [
  <h1>
    Search Results
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search hero"
          type="text"
          value="batman123"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No results found: 
        batman123
      </div>
      <ul
        className="list-group"
      />
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> Deberia mostrarse correctamente 1`] = `
Array [
  <h1>
    Search Results
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Search hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Please enter a search term
      </div>
      <ul
        className="list-group"
      />
    </div>
  </div>,
]
`;
